src/async_fuse/memfs/kv_engine/etcd_impl.rs

```diff
@@ -292,6 +625,92 @@ impl MetaTxn for EtcdTxn {
+    /// Old(reference to etcd client v3):
+    /// Create a key with prefix, and compare key (sorted by revision) with the key created by the same session.
+    /// If current key is the smallest, then the session is the master, and campaign success.
+    ///
+    /// New(For datenlord cache scenario):
+    /// A small batch of nodes in cluster, we just need a simple txn to get the master key.
+    /// The thundering herd is not the main problem, which will cause a bunch of etcd raft logs.
+    async fn campaign(
+        &self,
+        key: &KeyType,
+        val: &ValueType,
+        lease_id: i64,
+    ) -> (bool, Option<String>) {
+        let mut client = self.client.clone();
+
+        // Try to get the key, if key is existed
+        // We need to return the data from etcd,
+        // If the key is not existed, we need to set the key.
+        let txn = Txn::new()
+            .when(vec![Compare::create_revision(
+                key.to_string_key(),
+                CompareOp::Equal,
+                0,
+            )])
+            .and_then(vec![TxnOp::put(
+                key.to_string_key(),
+                serde_json::to_string(val).unwrap_or_else(|val| {
+                    panic!("failed to serialize value to json, value = {val:?}");
+                }),
+                Some(PutOptions::new().with_lease(lease_id)),
+            )])
+            .or_else(vec![TxnOp::get(
+                key.to_string_key(),
+                Some(GetOptions::new().with_serializable()),
+            )]);
+
+        let (campaign_status, responses) = match client.txn(txn).await {
+            Ok(resp) => {
+                // Check the txn branch is `then` or `else`
+                (resp.succeeded(), resp.op_responses())
+            }
+            Err(e) => {
+                error!("failed to campaign, error={e:?}");
+                (false, vec![])
+            }
+        };
+
+        // If the campaign is success, we need to check the response
+        if campaign_status {
+            // Check the response
+            if let Some(response) = responses.first() {
+                match *response {
+                    TxnOpResponse::Put(_) => {
+                        // we can directly return current data
+                        return (true, Some(key.to_string()));
+                    }
+                    TxnOpResponse::Get(ref resp) => {
+                        debug!("failed to campaign, return the existing key");
+                        if let Some(kv) = resp.kvs().first() {
+                            let item_key = match kv.key_str() {
+                                Ok(key) => key.to_owned(),
+                                Err(e) => {
+                                    error!("failed to get key from etcd watch event, error={e:?}");
+                                    return (false, None);
+                                }
+                            };
+
+                            return (true, Some(item_key));
```

Comment:
On get resp, should return item value, not the key (since key is always predefined).

Should add some error log for unexpected condition.

the function returning type can be defined as `Result<value,  Errtype>`.

Reply:
Ok, I have changed type to `DatenLordResult<(bool, Option<String>)>`.